app/models/compute_resources/foreman/model/vmware.rb

```diff
@@ -99,40 +99,56 @@ def clusters
       dc_clusters.map(&:full_path).sort
     end
 
+    # Params:
+    # +name+ identifier of the datastore - its name unique in given vCenter
+    def datastore(name)
+      cache.cache(:"datastore-#{name}") do
+        dc.datastores.get(name)
+      end
+    end
+
+    # ==== Options
+    #
+    # * +:cluster_id+ - Limits the datastores in response to the ones available to defined cluster
     def datastores(opts = {})
-      if opts[:storage_domain]
-        cache.cache(:"datastores-#{opts[:storage_domain]}") do
-          name_sort(dc.datastores.get(opts[:storage_domain]))
-        end
-      else
-        cache.cache(:datastores) do
-          name_sort(dc.datastores.all(:accessible => true))
+      cache.cache(cachekey_with_cluster(:datastores, opts[:cluster_id])) do
+        name_sort(dc.datastores(cluster: opts[:cluster_id]).all(:accessible => true))
+      end
+    end
+
+    def storage_pod(name)
+      cache.cache(:"storage_pod-#{name}") do
+        begin
+          dc.storage_pods.get(name)
+        rescue RbVmomi::VIM::InvalidArgument
+          {} # Return an empty storage pod hash if vsphere does not support the feature
         end
       end
     end
 
+    ##
+    # Lists storage_pods for datastore/cluster.
+    # TODO: fog-vsphere doesn't support cluster base filtering, so the cluser_id is useless for now.
+    # ==== Options
+    #
+    # * +:cluster_id+ - Limits the datastores in response to the ones available to defined cluster
     def storage_pods(opts = {})
       if opts[:storage_pod]
-        cache.cache(:"storage_pods-#{opts[:storage_pod]}") do
-          begin
-            dc.storage_pods.get(opts[:storage_pod])
-          rescue RbVmomi::VIM::InvalidArgument
-            {} # Return an empty storage pod hash if vsphere does not support the feature
-          end
-        end
+        Foreman::Deprecation.deprecation_warning('1.23', 'Use #storage_pod instead')
```

Comment:
I have been thinking about keeping the support in the model internally, but I think I was overdoing it. It is not really important method. On the api level, the support for the endpoint stays.

Reply:
yes, it is best to deprecate for 2 versions, and we can deprecate api methods but we don't actually ever remove them (or haven't so far...)

Reply:
I wrote a proposal to separate branch: https://github.com/ezr-ondrej/foreman/commit/4aa4ccfe15f70f0911143a1608eaeb61eddd067b. Just an idea it would be nice to have the deprecated endpoint separated. For future easier removal (given it is going to be probably at first foreman v3). But it got it's drawbacks, so it is just an idea. If you would like it, I can go further :)